src/gamma-wl.c

```diff
@@ -0,0 +1,372 @@
+		}
+	}
+	wl_display_flush(state->display);
+}
+
+static void
+wayland_free(wayland_state_t *state)
+{
+	int ret = 0;
+	/* Wait for the sync callback to destroy everything, otherwise
+	 * we could destroy the gamma control before gamma has been set */
+	while (state->callback && ret >= 0) {
+		ret = wl_display_dispatch(state->display);
+	}
+	if (state->callback) {
+		fprintf(stderr, _("Ignoring error on wayland connection while waiting to disconnect: %d\n"), ret);
+		wl_callback_destroy(state->callback);
+	}
+
+	for (int i = 0; i < state->num_outputs; ++i) {
+		struct output *output = &state->outputs[i];
+		if (output->gamma_control) {
+			zwlr_gamma_control_v1_destroy(output->gamma_control);
+			output->gamma_control = NULL;
+		}
+		wl_output_destroy(output->output);
+	}
+
+	if (state->gamma_control_manager) {
+		zwlr_gamma_control_manager_v1_destroy(state->gamma_control_manager);
+	}
+	if (state->registry) {
+		wl_registry_destroy(state->registry);
+	}
+	if (state->display) {
+		wl_display_disconnect(state->display);
+	}
+
+	free(state);
+}
+
+static void
+wayland_print_help(FILE *f)
+{
+	fputs(_("Adjust gamma ramps with a Wayland compositor.\n"), f);
+	fputs("\n", f);
+}
+
+static int
+wayland_set_option(wayland_state_t *state, const char *key, const char *value)
+{
+	return 0;
+}
+
+static void
+callback_done(void *data, struct wl_callback *cb, uint32_t t)
+{
+	wayland_state_t *state = data;
+	state->callback = NULL;
+	wl_callback_destroy(cb);
+}
+
+static const struct wl_callback_listener callback_listener = {
+	callback_done
+};
+
+static int
+wayland_set_temperature(wayland_state_t *state, const color_setting_t *setting)
+{
+	int ret = 0, roundtrip = 0;
+
+	/* We wait for the sync callback to throttle a bit and not send more
+	 * requests than the compositor can manage, otherwise we'd get disconnected.
+	 * This also allows us to dispatch other incoming events such as
+	 * wl_registry.global_remove. */
+	while (state->callback && ret >= 0) {
+		ret = wl_display_dispatch(state->display);
+	}
+	if (ret < 0) {
+		fprintf(stderr, _("The Wayland connection experienced a fatal error: %d\n"), ret);
+		return ret;
+	}
+
+	for (int i = 0; i < state->num_outputs; ++i) {
+		struct output *output = &state->outputs[i];
+		if (!output->gamma_control) {
+			output->gamma_control = zwlr_gamma_control_manager_v1_get_gamma_control(state->gamma_control_manager, output->output);
+			zwlr_gamma_control_v1_add_listener(output->gamma_control, &gamma_control_listener, output);
+			roundtrip = 1;
+		}
+	}
+	if (roundtrip) {
+		wl_display_roundtrip(state->display);
+	}
+
+	for (int i = 0; i < state->num_outputs; ++i) {
+		struct output *output = &state->outputs[i];
+		int size = output->gamma_size;
+		size_t ramp_bytes = size * sizeof(uint16_t);
+		size_t total_bytes = ramp_bytes * 3;
+
+		int fd = os_create_anonymous_file(total_bytes);
+		if (fd < 0) {
+			perror("os_create_anonymous_file");
+			return -1;
+		}
+
+		void *ptr = mmap(NULL, total_bytes,
```

Comment:
if gamma_size is 0 (i.e. when the output does not support setting gamma), the total_bytes will be zero and mmap will error with EINVAL.

https://gitlab.com/chinstrap/gammastep/-/commit/fdfec2bfb4ee8a44ae503a7c63a073453a17f176